src/workerd/api/html-rewriter.c++

```diff
@@ -410,6 +421,24 @@ void Rewriter::outputImpl(const char* buffer, size_t size) {
   }
 }
 
+void Rewriter::outputImpl(const char* buffer, size_t size) {
+  if (isPoisoned()) {
+    // Handlers disabled due to exception or running in a destructor.
+    return;
+  }
+
+  // Determine whether we should buffer the data.
+  int newBufferedLen = writeBufferPos + size;
+  if (newBufferedLen < writeBuffer.size()) {
+    // The new data fits in the write buffer, so write it there.
+    std::copy(buffer, buffer + size, writeBuffer.begin() + writeBufferPos);
+    writeBufferPos += size;
+    return;
+  }
+
+  writeFromBuffer(kj::heapArray(buffer, size));
```

Comment:
Since `writeFromBuffer()` is already copying its data to another buffer, I think this allocation/copy may be redundant... you might be able to pass a `kj::ArrayPtr` instead?